pkg/kallsyms/kallsyms_test.go

```diff
@@ -77,78 +84,153 @@ func TestRealKAllSyms(t *testing.T) {
+						Size:   8,
+					},
+					{
+						Type: &btf.Var{
+							Name: "socket_file_ops_addr",
+							Type: &btf.Int{Size: 8},
 						},
+						Offset: 8,
+						Size:   8,
 					},
 				},
-				Contents: []ebpf.MapKV{
-					{Key: uint32(0), Value: []byte{
-						0, 0, 0, 0, 0, 0, 0, 0,
-						0, 0, 0, 0, 0, 0, 0, 0,
-					}},
-				},
+			},
+			Contents: []ebpf.MapKV{
+				{Key: uint32(0), Value: []byte{
+					0, 0, 0, 0, 0, 0, 0, 0,
+					0, 0, 0, 0, 0, 0, 0, 0,
+				}},
 			},
 		},
-	}
+	},
+}
 
-	err := specUpdateAddresses(
-		kAllSymsFactory,
+func TestSpecRewriteConstants(t *testing.T) {
+	t.Cleanup(resetState)
+
+	var err error
+	kAllSymsStr := strings.Join([]string{
+		"0000000000000000 A fixed_percpu_data",
+		"ffffffffb4231f40 D bpf_prog_fops",
+		"ffffffffb43723e0 d socket_file_ops",
+	},
+		"\n")
+	kAllSymsReader := strings.NewReader(kAllSymsStr)
+	kAllSyms, err := NewKAllSymsFromReader(kAllSymsReader)
+	require.Nil(t, err, "NewKAllSymsFromReader failed: %v", err)
+	kSymbolResolver := newKAllSymsResolver()
+	kSymbolResolver.kAllSyms = kAllSyms
+
+	// Little endian representation of the addresses above:
+	bpfProgFopsAddr := []byte{0x40, 0x1f, 0x23, 0xb4, 0xff, 0xff, 0xff, 0xff}
+	socketFileOpsAddr := []byte{0xe0, 0x23, 0x37, 0xb4, 0xff, 0xff, 0xff, 0xff}
+
+	spec := specTest.Copy()
+
+	err = specUpdateAddresses(
+		[]symbolResolver{
+			kSymbolResolver,
+		},
 		spec,
 		[]string{"abcde_bad_name"},
 	)
-	require.ErrorIs(t, err, os.ErrNotExist, "specRewriteConstantsWithSymbolAddresses should have failed")
+	require.ErrorContainsf(t, err, "was not found in kallsyms", "specUpdateAddresses should have failed")
 
 	err = specUpdateAddresses(
-		kAllSymsFactory,
+		[]symbolResolver{
+			kSymbolResolver,
+		},
 		spec,
 		[]string{"bpf_prog_fops", "socket_file_ops"},
 	)
-	require.Nil(t, err, "specRewriteConstantsWithSymbolAddresses failed: %v", err)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
 
 	expectedContents := []byte{}
 	expectedContents = append(expectedContents, bpfProgFopsAddr...)
 	expectedContents = append(expectedContents, socketFileOpsAddr...)
 	contents := spec.Maps[".rodata"].Contents[0].Value
 	require.Equal(t, contents, expectedContents, "contents aren't equal")
 }
+
+func TestSpecRewriteConstantsRoot(t *testing.T) {
+	utilstest.RequireRoot(t)
+	utilstest.RequireFileContains(t, "/proc/kallsyms", "bpf_prog_fops")
+	utilstest.RequireFileContains(t, "/proc/kallsyms", "socket_file_ops")
+	t.Cleanup(resetState)
+
+	resetState()
+
+	spec1 := specTest.Copy()
+	err := specUpdateAddresses(
+		[]symbolResolver{
+			newKAllSymsResolver(),
+		},
+		spec1,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
+
+	resetState()
+
+	spec2 := specTest.Copy()
+	err = specUpdateAddresses(
+		[]symbolResolver{
+			newEbpfResolver(),
+		},
+		spec2,
+		[]string{"bpf_prog_fops", "socket_file_ops"},
+	)
+	require.Nil(t, err, "specUpdateAddresses failed: %v", err)
```

Comment:
This section of the test can work on kernels without `/proc/kallsyms`. Could you duplicate it in a new test `TestSpecRewriteConstantsWithEbpf` without the `utilstest.RequireFileContains`?
Of course, it won't check that the addresses have the correct value but it could at least check that they are non-zero and that `specUpdateAddresses()` does not return an error.

Reply:
Done, I see the test are still failing on the PR hopefully https://github.com/inspektor-gadget/inspektor-gadget/pull/2649 will help with it.